flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -949,7 +952,9 @@ public final void cleanUp(Throwable throwable) throws Exception {
         // disabled the interruptions or not.
         getCompletionFuture().exceptionally(unused -> null).join();
         // clean up everything we initialized
-        isRunning = false;
+        if (!isCanceled() && !isFailing()) {
```

Comment:
Technically, it is a race condition here. (as the same as several lines above in `failing` condition). We check the current state and only then set a new one. I don't think it is a big problem since we have finished the task already, but it is good to clarify this logic anyway.

Right now I see that we set the final state unconditionally in a couple of places while in a couple of other places, we check the condition that the state is not final yet. 

But I think actually that we have a pretty clear state machine here:
- init -> running
- running -> running(maybe)
- running -> canceled
- running -> failing
- running -> finished
So all other transfers should be ignored. So maybe we should come up with `compareAndSet` solution somehow?

@pnowojski , @rkhachatryan , WDYT?

P.S. Since the race condition is present in the current code as well we can implement it in the follow-up ticket rather than in this ticket.

Reply:
I would be cautious about any changes to the semantic here. Have you considered that in this PR you are flattening the matrix of valid combinations into the state machine that @akalash pointed out? For example before `cancelled` could be set `true` while `isRunning` was also `true`. Now that's no longer possible, and wouldn't such change in semantic cause some side-effects? 

Maybe it would be good to actually define the state machine that @akalash  proposed, but then also to go through all relevant getters (`isFailing()`/`isRunning()`/`isXXX()`), and check how are they being used and if they still make sense/are valid.

Reply:
I agree, FSM makes sense.

But I also agree, that all the combinations and changes are tricky.

So maybe if the purpose of the PR is to reduce `volatile` accesses, I'd rather move all three flags into the state object, and  reflect their exact same values there.
For example, replace `isRunning = false;` with `taskState = taskState.stopped();`  where `stopped()` returns a new state object with `running = false` inside.

On a side note, those `volatile` accesses aren't on the hot path so performance-wise it shouldn't change much.

Reply:
> On a side note, those volatile accesses aren't on the hot path so performance-wise it shouldn't change much

Yes, this doesn't affect performance. This ticket is about cleaning up the code and improving readability.

Reply:
I think that highly likely we can have FSM for this logic but I agree that it requires more research about its correctness.

So if we want to keep this ticket simple and save the logic that we have now. We can not have `failing` in the enum because we can have `failing` and `running` at the same time. Maybe we should have `TaskState` object with enum and `failing` flag inside. It doesn't look as clean as the solution in this PR  but at least it is cleaner than the current one and it keeps the same logic.

Reply:
Thanks all for the comments, I have been thinking about a State machine all along, just didnt want to introduce too many changes on such a fundamental class. 

@pnowojski  `canceled` and `running` can never be true at the same time because of: https://github.com/apache/flink/pull/21923/files#diff-0c5fe245445b932fa83fdaf5c4802dbe671b73e8d76f39058c6eaaaffd9639faL977-L978

However, we can not make the same assumption for `failing` and `running` as  @akalash mentioned.

I have added the transitions to cover the Task lifecycle in the latest commit but I am leaning towards adding the failing flag as part of the TaskState to keep the logic closer to the existing.